### PR TITLE
Use DateTimeImmutable instead of DateTimeInterface

### LINKED ISSUE #363: [RFC] Use DateTimeImmutable everywhere instead of DateTimeInterface
`DateTimeInterface` only contains the methods to format dates (`format`) and getters (`getTimestamp`, `getTimezone`, ...), as there are the only methods working exactly the same in `DateTime` and `DateTimeImmutable`. This makes the code a lot more complex when needing to manipulate dates (see the preSign logic for instance).
But allowing any DateTimeInterface has another drawback: the immutable value objects are not actually immutable. If I instantiate one of them and I provide a DateTime object, I can then actually mutate the date in this object.

I propose the following change:

- use `DateTimeImmutable` as type for all properties and return values in the library
- maybe keep accepting `DateTime` as input in constructors (either as a deprecated BC layer, or as a supported API), but convert the value to `DateTimeImmutable` (as done for string values)
- change argument types in internal APIs to use `DateTimeImmutable` (for the SignerV4 private APIs, I think that could actually already be done today, as it performs some conversion in public entrypoints anyway

What do you think @jderusse @Nyholm ?

### PR DESCRIPTION
**Pull request by Nyholm**
This will fix #363.

### CODE REVIEW

File: src/Core/src/Signer/SignerV4.php [hunk 1]
```diff
@@ -63,9 +63,8 @@ public function presign(Request $request, Credentials $credentials, RequestConte
     {
         if (null === $now = $context->getCurrentDate()) {
             $now = new \DateTimeImmutable();
-        } else {
-            $now = new \DateTimeImmutable($now->format(\DateTimeInterface::ATOM));
         }
+
```
Comment (jderusse) [inline src/Core/src/Signer/SignerV4.php]:
```suggestion
        $now = $context->getCurrentDate() ?? new \DateTimeImmutable();
```

File: src/Core/src/Signer/SignerV4.php [hunk 2]
```diff
@@ -77,9 +76,8 @@ public function sign(Request $request, Credentials $credentials, RequestContext
...
```
Comment (jderusse) [inline src/Core/src/Signer/SignerV4.php]:
```suggestion
        $now = $context->getCurrentDate() ?? new \DateTimeImmutable();
```

**Comment by Nyholm** [comment]
> I just wonder if we shouldn't allow \DateTimeInteface in input then convert the user input in \DateTimeImmutable

You mean Input objects, right? 
Because "inputs" like arguments to `presign` just caused these issues. 

Im not sure it is worth the trouble. Either way it can be added in the future without breaking BC.

File: src/Core/src/Signer/SignerV4.php (hunk 2 above)
Comment (Nyholm) [reply src/Core/src/Signer/SignerV4.php]: I could not apply these with the GUI

Screenshot 2020-03-27 at 00 47 45

**Pull request merged**